hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/ServerUtils.java

```diff
@@ -300,11 +304,189 @@ public static String getRemoteUserName() {
+      HddsProtos.NodeType nodeType) {
     LOG.warn("Storage directory for Ratis is not configured. It is a good " +
             "idea to map this to an SSD disk. Falling back to {}",
         HddsConfigKeys.OZONE_METADATA_DIRS);
     File metaDirPath = ServerUtils.getOzoneMetaDirPath(conf);
-    return (new File(metaDirPath, "ratis")).getPath();
+    
+    // Check for existing Ratis data from old versions for backward compatibility
+    String existingDir = findExistingRatisDirectory(metaDirPath, nodeType);
+    if (existingDir != null) {
+      return existingDir;
+    }
+    
+    // Use new component-specific location for new installations
+    String componentName = getComponentName(nodeType);
+    return (new File(metaDirPath, componentName + ".ratis")).getPath();
+  }
+
+  /**
+   * Checks for existing Ratis directories from previous versions for backward
+   * compatibility during upgrades.
+   *
+   * <p>Older versions of Ozone used different directory structures:
+   * <ul>
+   *   <li>Versions up to 2.0.0: Shared {@code <ozone.metadata.dirs>/ratis} for all components</li>
+   *   <li>Some SCM versions: Used {@code <ozone.metadata.dirs>/scm-ha}</li>
+   * </ul>
+   *
+   * @param metaDirPath The ozone metadata directory path
+   * @param nodeType Type of the node component
+   * @return Path to existing old Ratis directory if found, null otherwise
+   */
+  private static String findExistingRatisDirectory(File metaDirPath,
+      NodeType nodeType) {
+    // Check old Ratis directory locations in order of precedence
+    String[] oldRatisDirs = getOldRatisDirectoryNames(nodeType);
+    for (String dirName : oldRatisDirs) {
+      File oldRatisDir = new File(metaDirPath, dirName);
+      if (isNonEmptyDirectory(oldRatisDir)) {
+        LOG.info("Found existing Ratis directory at old location: {}. " +
+                "Using it for backward compatibility during upgrade.",
+            oldRatisDir.getPath());
+        return oldRatisDir.getPath();
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Returns array of old Ratis directory names to check, in order of precedence.
+   *
+   * @param nodeType Type of the node component
+   * @return Array of directory names to check
+   */
+  private static String[] getOldRatisDirectoryNames(NodeType nodeType) {
+    // Component-specific old locations take precedence over shared locations
+    if (nodeType == NodeType.SCM) {
+      return new String[]{"scm-ha", "ratis"};
+    }
+    return new String[]{"ratis"};
+  }
+
+  /**
+   * Converts NodeType enum to the component name string used for directory naming.
+   *
+   * @param nodeType Type of the node component
+   * @return Component name string (e.g., "om", "scm", "dn", "recon")
+   */
+  private static String getComponentName(NodeType nodeType) {
+    switch (nodeType) {
+    case OM:
+      return "om";
+    case SCM:
+      return "scm";
+    case DATANODE:
+      return "datanode";
```

Comment:
"dn" is used in other functions of Ozone. To keep it consistent, let's use dn instead of datanode.